### DISCUSSION TITLE
Same thing for morphMany relationship?

### DISCUSSION THREAD

**Discussion by danh65** [jorenvanhee/joren.co#5] [General]
This is a very nice piece of code. I'm wondering if it would be possible to do the same thing for morphMany relationship? Any thoughts on how to go about that?
Thanks, Dan

**Comment by jorenvanhee** [comment]
Laravel uses the following example in the docs for `morphMany`.

```
posts
    id - integer
    title - string
    body - text
 
videos
    id - integer
    title - string
    url - string
 
comments
    id - integer
    body - text
    commentable_id - integer
    commentable_type - string
```

I don't think there's anything stopping us from calling `upsert` (or `updateOrCreate`) on the `Comment` model. They only difference is that you would also have to specify the `commentable_type`. In my blog post we only specify the `recipe_id` during the `upsert`, in this case it will be the `commentable_id`.

You can use the [Laravel API docs](https://laravel.com/api/9.x/Illuminate/Database/Eloquent/Relations/MorphMany.html) to find useful methods on the `MorphMany` class. For instance, I see the following methods that might be useful:

- `getQualifiedForeignKeyName` Get the foreign key for the relationship.
- `getQualifiedMorphType` Get the foreign key "type" name.
- ...

I don't have time to fully build this, but I hope these tips help you to figure it out :)

**Comment by danh65** [comment]
Thank you for this info, it's very helpful! I'll see what I can come up with and maybe post it back here.
Thanks again!
Dan

**Comment by danh65** [comment]
I think I got it to work. I created another function that sets the correct attributes on the data and passes the data back as an array:
```
protected function prepareRecords(): array
  {
    $parentKey = $this->query->getParentKey();
    $morphClass = $this->query->getMorphClass();
    $morphName = $this->query->getQualifiedForeignKeyName();
    $morphType = $this->query->getQualifiedMorphType();
    return $this->records->map(function ($record) use($parentKey, $morphClass, $morphName, $morphType) {
      // Set $record['morph_id'] to parent key.
      $record[ $morphName ] = $parentKey;
      // Set $record['morph_type'] to parent class
      $record[ $morphType ] = $morphClass;
      // Set $record['id'] to null when missing.
      $record[$this->relatedKeyName] = array_key_exists($this->relatedKeyName, $record)
        ? $record[$this->relatedKeyName]
        : null;

      return $record;
    })->toArray();
  }
```

Then in the upsertRecords function, simply call that function:
``` 
protected function upsertRecords(): void
  {
    $values = $this->prepareRecords();
    (clone $this->query)->upsert(
      $values,
      [$this->relatedKeyName]
    );
  }
```

I made some changes to the construct method too, so the relatedKeyName is set once and used throughout the class:
```
private string $relatedKeyName;
  public function __construct(protected MorphMany $query, protected iterable $records)
  {
    // id (child table)
    $this->relatedKeyName = $query->getRelated()->getKeyName();
    $allowedRecordIds = $query->pluck($this->relatedKeyName);

    $this->records = collect($records)->filter(
      function ($record) use ($allowedRecordIds) {
        $id = $record[$this->relatedKeyName] ?? null;
        return $id === null || $allowedRecordIds->contains($id);
      }
    );
  }
```

The code for the AppServiceProvider is here:
```
        HasMany::macro('morphCreateUpdateOrDelete', function (iterable $records) {
          /** @var MorphMany $morphMany */
          $morphMany = $this;
          return (new MorphCreateUpdateOrDelete($morphMany, $records))();
        });
```

Hope this helps someone.
Dan
